walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,104 @@
+   * people that need to attend that meeting, and other events scheduled for the day. This
+   * algorithm is O(1) in terms of space complexity, requiring a boolean array of fixed size
+   * (24 * 60). It is O(mn) in time complexity, 'm' being the number of attendees required for the
+   * meeting and 'n' being the events of the day.
+   * @param events all the events scheduled for a given day, with their attendees and time range.
+   * @param request the request for a meeting, with their required attendees and minimum duration.
+   * @return the available time slots during the day where setting the meeting is possible.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    boolean[] unavailableMinutes = new boolean[24 * 60];
+    Collection<String> meetingAttendees = request.getAttendees();
+    long meetingDuration = request.getDuration();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      for (String meetingAttendee : meetingAttendees) {
+        if (eventAttendees.contains(meetingAttendee)) {
+          TimeRange eventTimeRange = event.getWhen();
+          addUnavailableMinutes(unavailableMinutes, eventTimeRange);
+          break;
+        }
+      }
+    }
+    
+    Collection<TimeRange> availableTimeSlots =
+        checkForAvailableTimeSlots(unavailableMinutes, meetingDuration);
+
+    return availableTimeSlots;
+  }
+
+  /**
+   * Adds the time range of a specific event to the array of unavailable minutes.
+   * @param unavailableMinutes the minutes where other events are happening.
+   * @param eventTimeRange the time range of a specific event to be added to unavailableMinutes.
+   */
+  private void addUnavailableMinutes(boolean[] unavailableMinutes, TimeRange eventTimeRange) {
+    if (eventTimeRange.start() < 0 || eventTimeRange.end() - 1 > 24 * 60 - 1) return;
+    for (int i = eventTimeRange.start(); i <= eventTimeRange.end() - 1; i++) {
+      unavailableMinutes[i] = true;
+    }
+  }
+
+  /**
+   * Creates a collection of available time slots with the information of unavailable minutes in the
+   * day and the minimum duration required.
+   * @param unavailableMinutes the minutes during the day where there can be no meetings.
+   * @param duration the minimum duration required for the requested meeting.
+   * @return a collection of time ranges where setting a meeting is possible.
+   */
+  private Collection<TimeRange> checkForAvailableTimeSlots(
+      boolean[] unavailableMinutes, long duration) {
+    // If duration required for the meeting is longer than a day, meeting is not possible.
+    if (duration > 24 * 60) {
+      return Collections.emptyList();
+    }
+
+    List<TimeRange> availableTimeSlots = new ArrayList<>();
+    boolean inTimeRange = false;
+    int timeRangeStart = 0;
+    int timeRangeEnd = 0;
```

Comment:
nit: consider changing these names to 'inAvailableTimeSlot', 'availableTimeSlotStart', etc., to make it more clear what time range these variables are referencing.

Reply:
Done!